enterprise/internal/campaigns/store_changeset_specs.go

```diff
@@ -380,3 +391,143 @@ func scanChangesetSpec(c *campaigns.ChangesetSpec, s scanner) error {
+// └───────────────────────────────────────┘    └───────────────────────────────┘
+// ┌───────────────────────────────────────┐    ┌───────────────────────────────────┐
+// │Changeset 5 | Repo C | | run-goimports │───▶│  Spec 3 | Repo C | run-goimports  │
+// └───────────────────────────────────────┘    └───────────────────────────────────┘
+//
+// Spec 1 should be attached to Changeset 1 and (possibly) update its title/body/diff.
+// Spec 2 should be attached to Changeset 2 and publish it on the code host.
+// Spec 3 should get a new Changeset, since its branch doesn't match Changeset 3's branch.
+// Spec 4 should be attached to Changeset 4, since it tracks PR #333 in Repo C.
+// Changeset 3 doesn't have a matching spec and should be detached from the campaign (and closed) (ChangesetSpec == 0, Changeset != 0).
+func (s *Store) GetRewirerMappings(ctx context.Context, opts GetRewirerMappingsOpts) (mappings campaigns.RewirerMappings, err error) {
+	q := getRewirerMappingsQuery(opts)
+
+	err = s.query(ctx, q, func(sc scanner) error {
+		var c campaigns.RewirerMapping
+		if err := sc.Scan(&c.ChangesetSpecID, &c.ChangesetID, &c.RepoID); err != nil {
+			return err
+		}
+		mappings = append(mappings, &c)
+		return nil
+	})
+	return mappings, err
+}
+
+func getRewirerMappingsQuery(opts GetRewirerMappingsOpts) *sqlf.Query {
+	return sqlf.Sprintf(
+		getRewirerMappingsQueryFmtstr,
+		opts.CampaignSpecID,
+		opts.CampaignID,
+		opts.CampaignID,
+		opts.CampaignSpecID,
+		opts.CampaignID,
+		opts.CampaignID,
+	)
+}
+
+var getRewirerMappingsQueryFmtstr = `
+-- source: enterprise/internal/campaigns/store_changeset_specs.go:GetRewirerMappings
+WITH
+	-- Fetch all changeset specs in the campaign spec that are of type 'existing'.
+	-- Match the entries to changesets in the target campaign by external ID and repo.
+	tracked_mappings AS (
+		SELECT changeset_specs.id AS changeset_spec_id, COALESCE(changesets.id, 0) AS changeset_id, changeset_specs.repo_id AS repo_id
+		FROM changeset_specs
+		LEFT JOIN changesets ON changesets.repo_id = changeset_specs.repo_id AND changesets.external_id = changeset_specs.spec->>'externalID'
+		INNER JOIN repo ON changeset_specs.repo_id = repo.id
+		WHERE
+		changeset_specs.campaign_spec_id = %s AND
+		changeset_specs.spec->>'externalID' IS NOT NULL AND changeset_specs.spec->>'externalID' != '' AND
+		repo.deleted_at IS NULL
+	),
+	-- Fetch all changeset specs in the campaign spec that are of type 'branch'.
+	-- Match the entries to changesets in the target campaign by head ref and repo.
+	branch_mappings AS (
+		SELECT changeset_specs.id AS changeset_spec_id, COALESCE(changesets.id, 0) AS changeset_id, changeset_specs.repo_id AS repo_id
+		FROM changeset_specs
+		LEFT JOIN changesets ON
+			changesets.repo_id = changeset_specs.repo_id AND
+			changesets.current_spec_id IS NOT NULL AND
+			((changesets.campaign_ids ? %s) OR changesets.owned_by_campaign_id = %s) AND
```

Comment:
I think it should be this:

```suggestion
			changesets.owned_by_campaign_id = %s AND
```

Otherwise this could match changesets that have a spec, but that are owned by another campaign, and that are merely attached-to/tracked-by _this_ campaign.

Reply:
very good! fixed and added a test